File: sdk-source/java/src/main/java/io/aegis/sdk/Aegis.java
package io.aegis.sdk;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/** Client for the Aegis Authentication API. Dependency-free (JDK 17+). */
public class Aegis implements AutoCloseable {
    private final String baseUrl;
    private final String appKey;
    private final String apiKey;
    private final String version;
    private final String channel;
    private final String hwid;
    private final int maxRetries;
    private final HttpClient http;
    private final Duration timeout;

    private volatile String sessionToken;
    private ScheduledExecutorService heartbeat;

    public Aegis(AegisOptions options) {
        if (options.baseUrl == null || options.baseUrl.isBlank())
            throw new AegisException("invalid_options", "baseUrl is required.", 0);
        if (options.appKey == null || options.appKey.isBlank())
            throw new AegisException("invalid_options", "appKey is required.", 0);

        this.baseUrl = options.baseUrl.replaceAll("/+$", "");
        this.appKey = options.appKey;
        this.apiKey = options.apiKey;
        this.version = options.version;
        this.channel = options.channel;
        this.hwid = options.hwid != null ? options.hwid : HardwareId.get();
        this.maxRetries = options.maxRetries;
        this.timeout = options.timeout;
        this.http = HttpClient.newBuilder().connectTimeout(options.timeout).build();
    }

    public String hwid() { return hwid; }

    public String sessionToken() { return sessionToken; }

    /** Restores a session token persisted by the host application. */
    public void useSession(String token) { this.sessionToken = token; }

    // ---------------- transport ----------------

    /** Calls any endpoint and returns its {@code data} payload. */
    public Map<String, Object> request(String endpoint, Map<String, Object> body) {
        String url = baseUrl + "/api/public/v1/" + endpoint.replaceAll("^/+|/+$", "");
        String payload = Json.write(body == null ? Map.of() : body);

        HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(url))
                .timeout(timeout)
                .header("content-type", "application/json")
                .header("x-app-key", appKey)
                .header("x-timestamp", String.valueOf(System.currentTimeMillis() / 1000))
                .header("user-agent", "aegis-java-sdk/1.0.0")
                .POST(HttpRequest.BodyPublishers.ofString(payload));
        if (apiKey != null) builder.header("x-api-key", apiKey);
        String token = sessionToken;
        if (token != null) builder.header("x-session-token", token);
        HttpRequest request = builder.build();

        Exception lastError = null;
        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
                if (response.statusCode() >= 500 && attempt < maxRetries) {
                    sleep(250L * (attempt + 1));
                    continue;
                }
                Map<String, Object> envelope = Json.parseObject(response.body());
                if (!Boolean.TRUE.equals(envelope.get("success"))) {
                    Object error = envelope.get("error");
                    String code = "server_error";
                    String message = "Request failed.";
                    if (error instanceof Map<?, ?> map) {
                        if (map.get("code") != null) code = String.valueOf(map.get("code"));
                        if (map.get("message") != null) message = String.valueOf(map.get("message"));
                    }
                    throw new AegisException(code, message, response.statusCode());
                }
                Object data = envelope.get("data");
                @SuppressWarnings("unchecked")
                Map<String, Object> result = data instanceof Map ? (Map<String, Object>) data : new LinkedHashMap<>();
                return result;
            } catch (AegisException exception) {
                throw exception;
            } catch (Exception exception) {
                lastError = exception;
                if (attempt < maxRetries) sleep(250L * (attempt + 1));
            }
        }
        throw new AegisException("network_error",
                lastError != null ? String.valueOf(lastError.getMessage()) : "Network request failed.", 0, lastError);
    }

    // ---------------- application ----------------

    /** Handshake. Call once before any other operation. */
    public Map<String, Object> init() {
        return request("init", Map.of("version", version));
    }

    public Map<String, Object> status() { return request("status", Map.of()); }

    public Map<String, Object> appData() { return request("app/data", Map.of()); }

    public Map<String, Object> checkVersion(String buildVersion) {
        return request("version/check", Map.of("version", buildVersion == null ? version : buildVersion, "channel", channel));
    }

    // ---------------- authentication ----------------

    public Map<String, Object> register(String username, String password, String email, String licenseKey) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("username", username);
        body.put("password", password);
        body.put("email", email);
        body.put("license_key", licenseKey);
        body.put("hwid", hwid);
        Map<String, Object> result = request("register", body);
        storeSession(result);
        return result;
    }

    public Map<String, Object> login(String username, String password) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("username", username);
        body.put("password", password);
        body.put("hwid", hwid);
        Map<String, Object> result = request("login", body);
        storeSession(result);
        return result;
    }

    public void logout() {
        stopHeartbeat();
        try {
            request("logout", Map.of());
        } finally {
            sessionToken = null;
        }
    }

    public Map<String, Object> heartbeat() { return request("heartbeat", Map.of()); }

    public Map<String, Object> checkSession() { return request("session/check", Map.of()); }

    /** True when a token exists and the server still accepts it. */
    public boolean isAuthenticated() {
        if (sessionToken == null) return false;
        try {
            return Boolean.TRUE.equals(checkSession().get("valid"));
        } catch (AegisException exception) {
            return false;
        }
    }

    public Map<String, Object> userData() { return request("user/data", Map.of()); }

    // ---------------- licensing ----------------

    public Map<String, Object> validateLicense(String licenseKey) {
        return request("license/validate", Map.of("license_key", licenseKey, "hwid", hwid));
    }

    public Map<String, Object> activateLicense(String licenseKey, String username) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("license_key", licenseKey);
        body.put("hwid", hwid);
        body.put("username", username);
        return request("license/activate", body);
    }

    // ---------------- variables ----------------

    public Map<String, Object> getVariables(String scope, String licenseKey) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("scope", scope == null ? "application" : scope);
        body.put("license_key", licenseKey);
        return request("variables/get", body);
    }

    public Map<String, Object> setVariable(String key, String value, String scope, String licenseKey) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("scope", scope == null ? "user" : scope);
        body.put("key", key);
        body.put("value", value);
        body.put("license_key", licenseKey);
        return request("variables/set", body);
    }

    public Map<String, Object> triggerWebhook(String event, Map<String, Object> payload) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("event", event);
        body.put("payload", payload == null ? Map.of() : payload);
        return request("webhook/trigger", body);
    }

    // ---------------- heartbeat loop ----------------

    /** Starts a background heartbeat until the session is revoked or stopped. */
    public synchronized void startHeartbeat(Duration interval, Consumer<String> onRevoked) {
        stopHeartbeat();
        long seconds = interval == null ? 60 : Math.max(5, interval.getSeconds());
        heartbeat = Executors.newSingleThreadScheduledExecutor(runnable -> {
            Thread thread = new Thread(runnable, "aegis-heartbeat");
            thread.setDaemon(true);
            return thread;
        });
        heartbeat.scheduleAtFixedRate(() -> {
            try {
                heartbeat();
            } catch (AegisException exception) {
                if (exception.isNetworkError()) return;
                sessionToken = null;
                if (onRevoked != null) onRevoked.accept(exception.getMessage());
                stopHeartbeat();
            }
        }, seconds, seconds, TimeUnit.SECONDS);
    }

    public synchronized void stopHeartbeat() {
        if (heartbeat != null) {
            heartbeat.shutdownNow();
            heartbeat = null;
        }
    }

    @Override
    public void close() {
        stopHeartbeat();
    }

    private void storeSession(Map<String, Object> result) {
        Object session = result.get("session");
        if (session instanceof Map<?, ?> map && map.get("token") != null) {
            sessionToken = String.valueOf(map.get("token"));
        }
    }

    private static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
        }
    }
}

File: sdk-source/kotlin/src/main/kotlin/io/aegis/sdk/Aegis.kt
package io.aegis.sdk

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.security.MessageDigest
import java.time.Duration
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/** Client configuration. */
data class AegisOptions(
    val baseUrl: String,
    val appKey: String,
    val apiKey: String? = null,
    val version: String = "1.0.0",
    val channel: String = "stable",
    val hwid: String? = null,
    val timeout: Duration = Duration.ofSeconds(20),
    val maxRetries: Int = 2,
)

/**
 * Client for the Aegis Authentication API.
 * JDK 17+, zero third-party dependencies. Thread-safe and [AutoCloseable].
 */
class Aegis(private val options: AegisOptions) : AutoCloseable {

    private val baseUrl = options.baseUrl.trimEnd('/')
    private val http: HttpClient = HttpClient.newBuilder().connectTimeout(options.timeout).build()
    private var heartbeatScheduler: ScheduledExecutorService? = null

    /** Machine identifier sent with authentication and licensing calls. */
    val hardwareId: String = options.hwid ?: hardwareId()

    /** Current session token, or null when signed out. */
    @Volatile
    var sessionToken: String? = null
        private set

    init {
        require(options.baseUrl.isNotBlank()) { "baseUrl is required." }
        require(options.appKey.isNotBlank()) { "appKey is required." }
    }

    companion object {
        const val SDK_VERSION = "1.0.0"

        /** Stable, non-reversible machine identifier. */
        fun hardwareId(): String {
            val facts = listOf(
                System.getProperty("os.name").orEmpty(),
                System.getProperty("os.arch").orEmpty(),
                System.getProperty("user.name").orEmpty(),
                runCatching { java.net.InetAddress.getLocalHost().hostName }.getOrDefault("unknown"),
            ).joinToString("|")
            return MessageDigest.getInstance("SHA-256").digest(facts.toByteArray())
                .joinToString("") { "%02x".format(it) }
        }
    }

    /** Restores a session token persisted by the host application. */
    fun useSession(token: String?) {
        sessionToken = token
    }

    /** Calls any endpoint and returns its `data` payload. */
    @Suppress("UNCHECKED_CAST")
    fun request(endpoint: String, body: Map<String, Any?> = emptyMap()): Map<String, Any?> {
        val payload = Json.encode(body.filterValues { it != null })
        val builder = HttpRequest.newBuilder(URI.create("$baseUrl/api/public/v1/${endpoint.trim('/')}"))
            .timeout(options.timeout)
            .header("content-type", "application/json")
            .header("user-agent", "aegis-kotlin-sdk/$SDK_VERSION")
            .header("x-app-key", options.appKey)
            .POST(HttpRequest.BodyPublishers.ofString(payload))
        options.apiKey?.let { builder.header("x-api-key", it) }
        sessionToken?.let { builder.header("x-session-token", it) }
        val httpRequest = builder.build()

        var lastFailure: Exception? = null
        for (attempt in 0..options.maxRetries) {
            try {
                val response = http.send(httpRequest, HttpResponse.BodyHandlers.ofString())
                if (response.statusCode() >= 500 && attempt < options.maxRetries) {
                    Thread.sleep(250L * (attempt + 1))
                    continue
                }
                val envelope = Json.decode(response.body().ifBlank { "{}" }) as? Map<String, Any?>
                    ?: throw AegisException("invalid_response", "Malformed API response.", response.statusCode())
                if (envelope.bool("success")) {
                    return envelope.obj("data")
                }
                val error = envelope.obj("error")
                throw AegisException(
                    error.str("code", "server_error"),
                    error.str("message", "Request failed."),
                    response.statusCode(),
                )
            } catch (error: AegisException) {
                throw error
            } catch (error: Exception) {
                lastFailure = error
                if (attempt < options.maxRetries) Thread.sleep(250L * (attempt + 1))
            }
        }
        throw AegisException("network_error", lastFailure?.message ?: "Network request failed.", 0, lastFailure)
    }

    // Application ---------------------------------------------------------

    /** Handshake. Call once before any other operation. */
    fun init(): Map<String, Any?> = request("init", mapOf("version" to options.version))

    fun status(): Map<String, Any?> = request("status")

    fun appData(): Map<String, Any?> = request("app/data")

    /** Download information published for this application. */
    fun downloads(): Map<String, Any?> = request("downloads")

    fun checkVersion(version: String? = null): Map<String, Any?> =
        request("version/check", mapOf("version" to (version ?: options.version), "channel" to options.channel))

    // Authentication ---------------------------------------------------------

    fun register(
        username: String,
        password: String,
        email: String? = null,
        licenseKey: String? = null,
    ): Map<String, Any?> = request(
        "register",
        mapOf(
            "username" to username, "password" to password, "email" to email,
            "license_key" to licenseKey, "hwid" to hardwareId,
        ),
    ).also(::storeSession)

    fun login(username: String, password: String): Map<String, Any?> = request(
        "login",
        mapOf("username" to username, "password" to password, "hwid" to hardwareId),
    ).also(::storeSession)

    fun logout() {
        try {
            request("logout")
        } finally {
            sessionToken = null
        }
    }

    fun heartbeat(): Map<String, Any?> = request("heartbeat")

    fun checkSession(): Map<String, Any?> = request("session/check")

    /** True when a token exists and the server still accepts it. */
    fun isAuthenticated(): Boolean {
        if (sessionToken == null) return false
        return runCatching { checkSession().bool("valid") }.getOrDefault(false)
    }

    fun userData(): Map<String, Any?> = request("user/data")

    // Licensing ---------------------------------------------------------

    fun validateLicense(licenseKey: String): Map<String, Any?> =
        request("license/validate", mapOf("license_key" to licenseKey, "hwid" to hardwareId))

    fun activateLicense(licenseKey: String, username: String? = null): Map<String, Any?> =
        request("license/activate", mapOf("license_key" to licenseKey, "hwid" to hardwareId, "username" to username))

    // Variables ---------------------------------------------------------

    fun getVariables(scope: String = "application", licenseKey: String? = null): Map<String, Any?> =
        request("variables/get", mapOf("scope" to scope, "license_key" to licenseKey))

    fun setVariable(
        key: String,
        value: String,
        scope: String = "user",
        licenseKey: String? = null,
    ): Map<String, Any?> = request(
        "variables/set",
        mapOf("scope" to scope, "key" to key, "value" to value, "license_key" to licenseKey),
    )

    fun triggerWebhook(event: String, payload: Map<String, Any?> = emptyMap()): Map<String, Any?> =
        request("webhook/trigger", mapOf("event" to event, "payload" to payload))

    // Sessions ---------------------------------------------------------

    /** Starts a daemon heartbeat; [onRevoked] fires once when the session dies. */
    fun startHeartbeat(interval: Duration = Duration.ofSeconds(60), onRevoked: (String) -> Unit = {}) {
        stopHeartbeat()
        val scheduler = Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "aegis-heartbeat").apply { isDaemon = true }
        }
        heartbeatScheduler = scheduler
        scheduler.scheduleAtFixedRate({
            try {
                heartbeat()
            } catch (error: AegisException) {
                stopHeartbeat()
                onRevoked(error.message ?: error.code)
            }
        }, interval.seconds, interval.seconds, TimeUnit.SECONDS)
    }

    fun stopHeartbeat() {
        heartbeatScheduler?.shutdownNow()
        heartbeatScheduler = null
    }

    override fun close() = stopHeartbeat()

    private fun storeSession(data: Map<String, Any?>) {
        val token = data.obj("session").str("token")
        if (token.isNotBlank()) sessionToken = token
    }
}